### PR TITLE
feat: business rules - create new module

### PR DESCRIPTION
**Pull request by pat-lewczuk**
## Summary

  This PR introduces a complete **Business Rules Engine** module to Open Mercato, enabling users to
  define, manage, and execute custom business logic without code changes. The engine provides a
  flexible, UI-driven approach to enforce validation rules, automate workflows, and implement complex
  business policies across entities.

  The Business Rules Engine allows users to:
  - Create conditional rules with a visual builder (no-code interface)
  - Execute automated actions based on rule outcomes
  - Group rules into reusable sets with execution ordering
  - Track rule execution history with detailed audit logs
  - Apply rules to any entity type at specific lifecycle events

### Core Engine (Backend)
  - **Rule Evaluation Engine** (`lib/rule-engine.ts`) - Orchestrates rule discovery, evaluation, and
  action execution
  - **Expression Evaluator** (`lib/expression-evaluator.ts`) - Evaluates complex conditions with 15+
  operators (=, >, <, IN, CONTAINS, REGEX, etc.)
  - **Action Executor** (`lib/action-executor.ts`) - Executes 10 action types (LOG, WEBHOOK, EMAIL,
  BLOCK, etc.)
  - **Value Resolver** (`lib/value-resolver.ts`) - Resolves field paths and special tokens ({{today}},
   {{user.email}}, etc.)
  - **Payload Validation** (`lib/payload-validation.ts`) - Server-side security checks (max nesting
  depth, field length limits)

  ### Database Schema
  - **4 Entities**: `BusinessRule`, `RuleExecutionLog`, `RuleSet`, `RuleSetMember`
  - **3 Migrations**: Initial schema, rule sets support, performance indexes
  - **JSONB Storage**: Flexible schema-less conditions and actions with PostgreSQL JSON operators
  - **Multi-tenant Isolation**: All entities scoped by `tenant_id` and `organization_id`
  - **Soft Delete Support**: Queries exclude `deleted_at IS NOT NULL` by default

  ### API Endpoints (13 Routes)
  - **Rules CRUD**: GET/POST/PATCH/DELETE `/api/business_rules/rules`
  - **Rule Execution**: POST `/api/business_rules/execute` - Execute rules for entity/event
  - **Execution Logs**: GET `/api/business_rules/logs` - Query audit trail with filtering
  - **Rule Sets**: Full CRUD for grouping and managing related rules
  - **Set Members**: Manage rules within sets with execution sequencing

  ### Admin UI (8 Pages)
  - **Rules List** (`backend/rules/page.tsx`) - DataGrid with search, filters, bulk actions
  - **Create/Edit Rule** (`backend/rules/[id]/page.tsx`) - Form with visual condition/action builders
  - **Rule Sets List** (`backend/sets/page.tsx`) - Manage rule groupings
  - **Set Members** (`backend/sets/[id]/RuleSetMembers.tsx`) - Drag-and-drop rule ordering
  - **Execution Logs** (`backend/logs/page.tsx`) - Audit trail with advanced filtering

  ### UI Components
  - **ConditionBuilder** - Visual no-code condition editor with AND/OR grouping
  - **ActionBuilder** - Configure actions with type-specific forms
  - **Field Validation** - Real-time validation with helpful error messages
  - **CrudForm Enhancements** - Auto-suggestions for entity types, event types, field paths

  ### Security & Performance
  - **RBAC Integration**: 7 feature permissions (`business_rules.rules.view`, `.create`, `.edit`,
  `.delete`, `.execute`, `.logs.view`, `.sets.manage`)
  - **SQL Injection Prevention**: ORM-based queries with parameterization
  - **ReDoS Protection**: Regex validation, timeouts, complexity limits
  - **Sensitive Data Sanitization**: 13 patterns redacted in logs (password, token, ssn, etc.)
  - **Execution Limits**: Max 100 rules per execution, 10-level nesting depth, 30s per-rule timeout
  - **Strategic Indexes**: Entity type + event type, priority ordering, tenant isolation

  ### Documentation
  - **Architecture Guide** (`docs/framework/business-rules/architecture.mdx`) - 541 lines covering
  design principles, data layer, services, security
  - **Services Reference** (`docs/framework/business-rules/services.mdx`) - 676 lines with API
  documentation for all core services
  - **Extending Guide** (`docs/framework/business-rules/extending.mdx`) - 442 lines on custom
  operators, actions, value resolvers
  - **User Guide** (7 docs) - Creating rules, rule types, conditions, actions, rule sets, execution
  logs
  - **API Documentation** (5 docs) - Complete OpenAPI specs for all endpoints
  - **Tutorial** - Material availability use case with step-by-step implementation

  ### Testing
  - **279 Passing Tests** across all components
  - **Unit Tests**: Expression evaluator, action executor, value resolver
  - **API Tests**: All 13 endpoints with success/error cases
  - **Coverage**: Core services, edge cases, error handling, security validations

  ### Internationalization
  - **English & Polish** translations for all UI strings
  - Rule type names, operator labels, action descriptions
  - Error messages, validation feedback, help text

## Testing

```bash
  # Run all business rules tests
  yarn test --testMatch="**/business_rules/**/*.test.ts"

  # Run rule engine unit tests
  yarn test --testMatch="**/rule-engine.test.ts"

  # Generate and apply migrations
  yarn db:generate
  yarn db:migrate

  # Initialize with business rules features
  yarn mercato init

  # Start development server
  yarn dev
```

## Checklist

- [x] This pull request targets `develop`.
- [x] I have read and accept the Open Mercato Contributor License Agreement (see `docs/cla.md`).
- [x] I updated documentation, locales, or generators if the change requires it.
- [x] I added or adjusted tests that cover the change.

## Linked issues

N/A

### CODE REVIEW

File: docs/docs/api/business-rules/execute.mdx [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+### Pre-Save Validation
+
+Execute rules before database writes:
+
+```typescript
+app.post('/api/orders', async (req, res) => {
+  const orderData = req.body
+
+  // Execute validation rules
+  const ruleResult = await executeRules({
```
Comment (pkarw) [inline docs/docs/api/business-rules/execute.mdx:412]: @pat-lewczuk are we going to add those in all `commands` and routes (maybe it's already there in this PR - i'm reading it from top down)

File: docs/docs/user-guide/business-rules/actions.mdx [hunk 2]
```diff
@@ -0,0 +1,389 @@
+---
```
Comment (pkarw) [inline docs/docs/user-guide/business-rules/actions.mdx:1]: @pat-lewczuk add UI screenshots please, we put screenshots in to the `docs/static/screenshots` folder

File: packages/core/src/modules/auth/lib/setup-app.ts [hunk 3]
```diff
@@ -310,6 +310,7 @@ async function ensureDefaultRoleAcls(
       'api_keys.*',
       'perspectives.use',
       'perspectives.role_defaults',
+      'business_rules.*',
```
Comment (pkarw) [inline packages/core/src/modules/auth/lib/setup-app.ts:319]: @pat-lewczuk make the `yarn mercato init` adding this feature to the `admin` role - its the `mercato auth seed-roles` command which makes it happen - so we simply need to extend it

File: packages/core/src/modules/business_rules/api/rules/route.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
+import { NextResponse } from 'next/server'
```
Comment (pkarw) [inline packages/core/src/modules/business_rules/api/rules/route.ts:1]: @pat-lewczuk in other places we're using 2 useful libraries:
- crud factory helpers for making re-usable API's: [an example from customers people api](https://github.com/open-mercato/open-mercato/blob/0de042695dd2a6df3cb03350122b834df524bc83/packages/core/src/modules/customers/api/people/route.ts#L52)
- commands pattern - which makes the write operations undoable - [an example from customers people commands](https://github.com/open-mercato/open-mercato/blob/0de042695dd2a6df3cb03350122b834df524bc83/packages/core/src/modules/customers/commands/people.ts#L392)

Again - I guess we don't have to refactor it here which could increase the complexity probably - but I just wanted to point it out more like for the future because it could be very useful, especially when dealing with `custom fields` (bc. commands and crud-factory are doing quite OK with it already)

File: packages/core/src/modules/business_rules/api/rules/route.ts [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    entityType: 'entityType',
+    priority: 'priority',
+    createdAt: 'createdAt',
+    updatedAt: 'updatedAt',
+  }
+
+  const orderByField = sortField && sortFieldMap[sortField] ? sortFieldMap[sortField] : 'priority'
+  const orderBy = { [orderByField]: sortDir, ruleId: 'asc' as const }
+
+  const [rows, count] = await em.findAndCount(
```
Comment (pkarw) [inline packages/core/src/modules/business_rules/api/rules/route.ts:136]: @pat-lewczuk we can skip applying the commands and crud-factory, but I guess using the `query-engine` could be very usefull here and in other places you're doing database queries as it's dealing with pagination (and custom fields if needed)

File: packages/core/src/modules/business_rules/lib/rule-engine.ts [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+      return false
+    }
+    return true
+  })
+}
+
+/**
+ * Sensitive field patterns to exclude from logs
+ */
+const SENSITIVE_FIELD_PATTERNS = [
```
Comment (pkarw) [inline packages/core/src/modules/business_rules/lib/rule-engine.ts:410]: let's make this filtering a reusable pattern in the core module - plus add info about it to `agents.md` so we'll be using it in the next developments @pat-lewczuk

**Comment by pkarw** [comment]
some more suggestions - here in the comments

1. Add a proper icons in the `page.meta.ts`:

image

**Comment by pkarw** [comment]
2. Move the whole section `Business Rules` higher in the sidebar - you can do it by setting the `pageOrder: 110` in the `page.meta`

**Comment by pat-lewczuk** [comment]
> Sure, let's make it in second Pr. I didn't get the `eventBus` integration part so that why the question; makes complete sense now
> 
> I think it would be great to add a `notifications` module where for example this Info action generates a notification.
> 
> If you're in the app - you see it right away.
> 
> If you're not - you have it somewhere in the toopbar like for example notifications in the bank / finance apps.
> 
> They should've got the actions as well so you can accept something etc
> 
> What do you think?
> 
> Pushing notification can work on a specific event from the eventBus so notifications module is just listening to it - by having it this way they gonna work if you're in the using or not, using API or not and whatsoever

Yes I will need to have it for workflow module as well, I will push some notification e.g. you need to approve something I will send you a notification should to do so :).

Let's talk about it and we will decide where to put it on road map. So exciting !

**Review by pkarw — CHANGES_REQUESTED** [review]
@pat-lewczuk great job! We can merge it. I'd only suggest to align the UI layouts (see screenshots attached) to be coherent with the rest of the app. After fixing the forms I'm happy to merge it in

File: docs/docs/api/business-rules/execute.mdx (hunk 1 above)
Comment (pat-lewczuk) [reply docs/docs/api/business-rules/execute.mdx:412]: Yes, this is for now something that is not implemented yet.  I will for sure add this as a integration work in next step. Sorry, you have right that it is currently misleading.

File: packages/core/src/modules/business_rules/lib/rule-engine.ts (hunk 6 above)
Comment (pat-lewczuk) [reply packages/core/src/modules/business_rules/lib/rule-engine.ts:410]: @pkarw you are talking about sensitive_filed_patters or filter by effective date?

File: packages/core/src/modules/auth/lib/setup-app.ts (hunk 3 above)
Comment (pat-lewczuk) [reply packages/core/src/modules/auth/lib/setup-app.ts:319]: @pkarw Could you please explain me what need to be done here? I see seed-roles command only add roles not add any permissions to that roles. Should I extend current function of seed-roles command to provide possibility to add permissions to roles as well in that command?

Comment (pkarw) [reply packages/core/src/modules/auth/lib/setup-app.ts:319]: It's ok! Will work in this form - it's going to add permissions to all features on `business_rules` to the seeded roles so to the admin user as a `consequence` 

Thanks!

File: packages/core/src/modules/business_rules/lib/rule-engine.ts (hunk 6 above)
Comment (pkarw) [reply packages/core/src/modules/business_rules/lib/rule-engine.ts:410]: Yes, exactly this - I think it can be useful application wide

File: packages/core/src/modules/business_rules/api/rules/route.ts (hunk 4 above)
Comment (pat-lewczuk) [reply packages/core/src/modules/business_rules/api/rules/route.ts:1]: I read through this examples and I think now I'm understand how it works. Thanks for pointing this to me.

**Review by pkarw — APPROVED** [review]
All good. Merging in.
